velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -202,310 +195,211 @@ inline static TResult addLarge(
-    if (args[0]->isConstantEncoding() && args[1]->isFlatEncoding()) {
-      // Fast path for (const, flat).
-      auto constant = args[0]->asUnchecked<SimpleVector<A>>()->valueAt(0);
-      auto flatValues = args[1]->asUnchecked<FlatVector<B>>();
-      auto rawValues = flatValues->mutableRawValues();
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            constant,
-            rawValues[row],
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
-    } else if (args[0]->isFlatEncoding() && args[1]->isConstantEncoding()) {
-      // Fast path for (flat, const).
-      auto flatValues = args[0]->asUnchecked<FlatVector<A>>();
-      auto constant = args[1]->asUnchecked<SimpleVector<B>>()->valueAt(0);
-      auto rawValues = flatValues->mutableRawValues();
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            rawValues[row],
-            constant,
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
-    } else if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
-      // Fast path for (flat, flat).
-      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
-      auto rawA = flatA->mutableRawValues();
-      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
-      auto rawB = flatB->mutableRawValues();
-
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            rawA[row],
-            rawB[row],
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
+// Add the value a and b, set the result r. The caller should check overflow
+// flag first.
+template <typename TResult, typename A, typename B>
+inline void applyAdd(
```

Comment:
Because multiply function will set the overflow flag
```
  multiply(int64_t a, int64_t b, bool& overflow) {
    int64_t value;
    overflow = __builtin_mul_overflow(a, b, &value);
    return value;
  }
```